src/EFCore.PG/Extensions/NpgsqlIndexBuilderExtensions.cs

```diff
@@ -47,5 +47,24 @@ public static IndexBuilder ForNpgsqlHasOperators([NotNull] this IndexBuilder ind
 
             return indexBuilder;
         }
+
+        /// <summary>
+        /// The PostgreSQL index sort options to be used.
+        /// </summary>
+        /// <remarks>
+        /// https://www.postgresql.org/docs/current/static/indexes-ordering.html
+        /// </remarks>
+        /// <param name="indexBuilder"> The builder for the index being configured. </param>
+        /// <param name="sortOptions"> The sort options to use for each column. </param>
+        /// <returns> A builder to further configure the index. </returns>
+        public static IndexBuilder ForNpgsqlHasSortOptions([NotNull] this IndexBuilder indexBuilder, [CanBeNull] params string[] sortOptions)
+        {
+            Check.NotNull(indexBuilder, nameof(indexBuilder));
+            Check.NullButNotEmpty(sortOptions, nameof(sortOptions));
+
+            indexBuilder.Metadata.Npgsql().SortOptions = sortOptions;
```

Comment:
Should we do some form of validation here? Otherwise these get passed right on through to the SQL below. First thought is typos, second thought is a stray semicolon...

Reply:
If you can come up with some kind of validation, then I'd be happy to hear about it. [The docs about index ordering](https://www.postgresql.org/docs/current/indexes-ordering.html) only mentions B-tree:

> You can adjust the ordering of a B-tree index by including the options ASC, DESC, NULLS FIRST, and/or NULLS LAST when creating the index

And then [the operator classes docs](https://www.postgresql.org/docs/current/indexes-opclass.html) says

> The operator class determines the basic sort ordering (which can then be modified by adding sort options COLLATE, ASC/DESC and/or NULLS FIRST/NULLS LAST).

Does this mean that other index types or operator classes can have other sort options? In that case, how do we validate that?

If the only valid values are `COLLATE`, `ASC`, `DESC`, `NULLS FIRST` and `NULLS LAST` (and any number of combination of them), I guess we could do some validation, but my impression was that it would be "too much".

The docs also mention that

> Obviously, indexes with non-default sort orderings are a fairly specialized feature, but sometimes they can produce tremendous speedups for certain queries. Whether it's worth maintaining such an index depends on how often you use queries that require a special sort ordering.

So I'm not really worried that a lot of people will use this API incorrectly.

Reply:
If we add validation, I think it would be fine to just implement the standard options for now, and wait for more exotic options to be requested.

I was thinking it might make sense to implement it with a flag enum:

```c#
[Flag]
public enum SortOptions
{
    None                                       = 0,
    [Display(Name = "COLLATE")]     Collate    = 1 << 0,
    [Display(Name = "ASC")]         Ascending  = 1 << 1,
    [Display(Name = "DESC")]        Descending = 1 << 2,
    [Display(Name = "NULLS FIRST")] NullsFirst = 1 << 3,
    [Display(Name = "NULLS LAST")]  NullsLast  = 1 << 4
}

public static IndexBuilder ForNpgsqlHasSortOptions(
    [NotNull] this IndexBuilder indexBuilder, 
    [CanBeNull] params SortOptions[] sortOptions) { ... }

builder.ForNpgsqlHasSortOptions(SortOptions.Ascending | SortOptions.NullsLast,);
```

That wouldn't stop someone from combining contradictory states (e.g. `ASC DESC`), but it would restrict the values that one could provide. Then again, it might just be overkill to perform validation here.

@roji Thoughts?

Reply:
OK, I have an idea. What if we keep the `SortOptions` property as `string[]` and just change the extension method to accept `params SortOptions[] sortOptions)` and do the conversion in there? That way, you can always fall back to setting the `string[]` directly. The extension method is just a convenience.

Reply:
I'm definitely for providing a more structured API, rather than a string-based one. One of the major points of using EF Core is that you're no longer writing arbitrary strings, but rather using well-crafted rich APIs. If users need/want to write arbitrary SQL in string, they can always drop down to raw SQL in their migrations.

I'd actually expose the index options via several methods, such as `ForNpgsqlHasOrderDescending(bool)`, `ForNpgsqlHasNullOrderingReversed(bool)` (the names may be a bit long. Each would modify its own annotation on the index, no need to mix up things which aren't related.

What do you guys think? If so this would have quite a big impact on this PR so I won't continue reviewing for now.

Reply:
Interesting idea. That might even fit in better with the rest of the fluent API. Plus, it would avoid introducing an enum for one method.

Reply:
> I'd actually expose the index options via several methods, such as `ForNpgsqlHasOrderDescending(bool)`, `ForNpgsqlHasNullOrderingReversed(bool)` (the names may be a bit long. Each would modify its own annotation on the index, no need to mix up things which aren't related.

How is this supposed to work with multiple columns?

Reply:
@khellang good point.

First, it could make sense to introduce single-parameter method overloads for single-column indices, which seem to be the common case. I'd expect these to cause an informative exception to be thrown if these are used if a multi-column index (directing the user to use the multi-column overload instead).

For the multi-column overloads, it seems that having `params bool[]` is a good option, no? Keeping the methods different (as opposed to having just one method accepting enums) makes sense to me, because (a) each could/should modify a distinct annotation on the index, and (b) null ordering, asc/desc and collation seem to be very unrelated concerns which seem to justify different methods.

What do you think?

Reply:
@roji I pushed an update. What do you think about something like that?

Reply:
Looks nice! The only thing I'm not sure about is having 3 values there (true/false/null).

For ASC/DESC, ASC is documented to be the default, so it seems like we could just have true/false (where false simply means don't render anything.

NULLS FIRST/LAST is more tricky, because the default actually depends on ASC/DESC, so a null option seems to make more sense there.

Reply:
Yeah, I just went with `null` to mean "just give me the default", so you don't have to worry about which option actually is the default. It's consistent with operators and method (btree) as well.

Reply:
OK, although I find it weird that for ASC/DESC null and false mean exactly the same thing - seems better to leave just true/false. For NULLS FIRST/LAST the situation is different and null does seem to make sense.

Reply:
> OK, although I find it weird that for ASC/DESC null and false mean exactly the same thing - seems better to leave just true/false.

Are you sure this is always true? Why would you special case this compared to all the other annotations which allow null?

The Postgres docs says:

> By default, B-tree indexes store their entries in ascending order with nulls last.

What about non-B-tree indexes? Do they have the same default?

I really think we should leave an option for "just use the default" in addition to true/false.

Reply:
Wouldn't it be better to an enum like `SortOrder` from SqlClient?

Reply:
>> OK, although I find it weird that for ASC/DESC null and false mean exactly the same thing - seems better to leave just true/false.
>
> Are you sure this is always true? Why would you special case this compared to all the other annotations which allow null?

I just looked at [the docs here](https://www.postgresql.org/docs/current/sql-createindex.html):

> `ASC` Specifies ascending sort order (which is the default).

I agree that if there's any doubt about the default here, then null should be a good option - I'm OK with whatever you decide here.

> Wouldn't it be better to an enum like `SortOrder` from SqlClient?

That's also possible - we'd still need to have an "unspecified" option thought, just like the SqlClient version (to avoid rendering anything in SQL). I agree that this is a better API than bools (where we can get confused about the meaning). If we do this, we would do the same for nulls first/nulls last as well.

@khellang does this make sense to you?

PS I know this conversation is getting a bit long on a relatively small point :) I think we're almost there...

Reply:
I'm fine with whatever. I'm generally not a fan of `null`, but at least it was consistent with existing metadata. If we have an enum value of `Unspecified` that would achieve the same thing, that's fine 👍

Reply:
OK, so let's go with two enums (SortOrder, NullOrdering?), each of which has an Unspecified value of 0 (for sane default).